src/litserve/specs/openai_embedding.py

```diff
@@ -0,0 +1,155 @@
+        return request.input if isinstance(request.input, list) else [request.input]
+
+    def predict(self, x)-> List[List[float]]:
+        return np.random.rand(len(x), 768).tolist()
+
+    def encode_response(self, output)-> dict:
+        return {"embeddings": output}
+```
+"""
+
+RESPONSE_VALIDATION_MSG = LITAPI_VALIDATION_MSG.split("\n", 2)[-1]
+
+
+class OpenAIEmbeddingSpec(LitSpec):
+    def __init__(self):
+        super().__init__()
+        # register the endpoint
+        self.add_endpoint("/v1/embeddings", self.embeddings, ["POST"])
+        self.add_endpoint("/v1/embeddings", self.options_embeddings, ["GET"])
+
+    def setup(self, server: "LitServer"):  # noqa: F821
+        from litserve import LitAPI
+
+        super().setup(server)
+
+        lit_api = self._server.lit_api
+        if inspect.isgeneratorfunction(lit_api.predict):
+            raise ValueError(LITAPI_VALIDATION_MSG.format("predict is a generator"))
+
+        is_encode_response_original = lit_api.encode_response.__code__ is LitAPI.encode_response.__code__
+        if not is_encode_response_original and inspect.isgeneratorfunction(lit_api.encode_response):
+            raise ValueError(LITAPI_VALIDATION_MSG.format("encode_response is a generator"))
+
+        print("OpenAI Embedding Spec is ready.")
+
+    def decode_request(self, request: EmbeddingRequest, context_kwargs: Optional[dict] = None) -> List[str]:
+        return request.get_input_as_list()
+
+    def encode_response(self, output: List[List[float]], context_kwargs: Optional[dict] = None) -> dict:
+        return {
+            "embeddings": output,
+            "prompt_tokens": context_kwargs.get("prompt_tokens", 0),
+            "total_tokens": context_kwargs.get("total_tokens", 0),
+        }
+
+    def validate_response(self, response: dict) -> None:
+        if not isinstance(response, dict):
+            raise ValueError(RESPONSE_VALIDATION_MSG.format("response is not a dictionary"))
+        if "embeddings" not in response:
+            raise ValueError(RESPONSE_VALIDATION_MSG.format("embeddings key is missing in response"))
+
+    async def embeddings(self, request: EmbeddingRequest):
+        response_queue_id = self.response_queue_id
+        logger.debug("Received embedding request: %s", request)
+        uid = uuid.uuid4()
+        event = asyncio.Event()
+        self._server.response_buffer[uid] = event
+
+        self._server.request_queue.put_nowait((response_queue_id, uid, time.monotonic(), request.model_copy()))
+        await event.wait()
+
+        response, status = self._server.response_buffer.pop(uid)
+
+        if status == LitAPIStatus.ERROR:
+            raise response
+
+        logger.debug(response)
+
+        # TODO: Validate response and also confirm if UsageInfo should be default or not, as None is also a valid value
+        # maybe move this validate to setup for early validation
+        self.validate_response(response)
```

Comment:
Since you are returning a dict, we don't have a lot of structure to work on, so moving to setup is not really an option. It's also a good opportunity to produce a good error message.